making extestionAPI be available as a Promise to simplify bootstapping + fixed raced condition for query transformers

twop opened the pull request:
done: 
- endpoint pair is now obtained by Promise vs Observable
- simplified bootstapping of extension api
- fixed query transformer race condition by explicitly waiting for extension api promise

left 
- fix failing tests. I need to properly stub it, naive attempt failed.

inspired by https://github.com/sourcegraph/sourcegraph/pull/11581/files

felixfbecker commented on shared/src/platform/context.ts:
```diff
@@ -18,6 +18,11 @@ export interface EndpointPair {
 }
 
+export interface ClosablePair {
+    pair: EndpointPair
```
I'd rename this to `endpoints`, because it's more important _what_ this is a pair _of_ (semantic), as opposed to that it is a pair of _something_ (type). You actually renamed this in one of the destructurings to `endpoints` so that would avoid the rename there.
```suggestion
    endpoints: EndpointPair
```

felixfbecker commented on shared/src/platform/context.ts:
```diff
@@ -18,6 +18,11 @@ export interface EndpointPair {
 
+export interface ClosablePair {
+    pair: EndpointPair
+    close: () => void
```
I'd make this a Subscription like [@lguychard had suggested](https://sourcegraph.slack.com/archives/CMT39K56Z/p1591355689267900?thread_ts=1591310354.265900&cid=CMT39K56Z) for the [reasons I explained in the same thread](https://sourcegraph.slack.com/archives/CMT39K56Z/p1591702099293700?thread_ts=1591310354.265900&cid=CMT39K56Z)
```suggestion
    /** Closes the endpoint pair. */
    subscription: Unsubscribable
```

felixfbecker commented on shared/src/platform/context.ts:
```diff
@@ -18,6 +18,11 @@ export interface EndpointPair {
     expose: Endpoint
 }
 
+export interface ClosablePair {
```
(see comment below)
```suggestion
export interface ClosableEndpointPair {
```

felixfbecker commented on shared/src/api/client/connection.ts:
```diff
@@ -145,11 +148,11 @@ export async function createExtensionHostClientConnection(
-    // until we untangle the bootstrap sequence of the extension host
-    services.queryTransformer.transformQuery = query => wrapRemoteObservable(proxy.transformSearchQuery(query))
+    // // TODO (simon) this is ugly mutation but not much can be done here
+    // // until we untangle the bootstrap sequence of the extension host
+    // services.queryTransformer.transformQuery = query => wrapRemoteObservable(proxy.transformSearchQuery(query))
 
-    subscription.add(() => (services.queryTransformer.transformQuery = undefined))
+    // subscription.add(() => (services.queryTransformer.transformQuery = undefined))
```
Can this be removed, or is there a reason it's only commented out?

twop replied:
done

felixfbecker commented on shared/src/api/extension/worker.ts:
```diff
@@ -21,10 +20,7 @@ export function createExtensionHostWorker(): { worker: ExtensionHostWorker; clie
-    })
+export function createExtensionHost(): Promise<ClosablePair> {
+    const { clientEndpoints, worker } = createExtensionHostWorker()
+    return Promise.resolve({ pair: clientEndpoints, close: () => worker.terminate() })
```
Could this function just be synchronous? (note that the Observable before was synchronous, but now it's async since Promises are always async)

twop replied:
Nope :( because of iframe init being async

twop replied:
Our platform context interfaces exposes a promise (used to be an async observable) because of iframe instantiation. 
But I think in that case we can lift that concern to platform context.

felixfbecker commented on shared/src/api/client/services.ts:
```diff
@@ -52,7 +51,7 @@ export class Services {
     public readonly textDocumentHover = new TextDocumentHoverProviderRegistry()
     public readonly textDocumentDecoration = new TextDocumentDecorationProviderRegistry()
-    public readonly queryTransformer: { transformQuery?: (query: string) => Observable<string> } = {}
+    // public readonly queryTransformer: { transformQuery?: (query: string) => Observable<string> } = {}
```
Can this be removed?

twop replied:
done

felixfbecker commented on browser/src/shared/platform/extensionHost.ts:
```diff
@@ -41,25 +41,27 @@ function createInPageExtensionHost({
                     Object.values(clientEndpoints)
                 )
-                subscriber.next(workerEndpoints)
+                resolve({
+                    pair: workerEndpoints,
+                    close: () => {
+                        clientEndpoints.proxy.close()
+                        clientEndpoints.expose.close()
+                        frame.remove()
+                    },
+                })
```
Something that is confusing me here is that with a `ClosableEndpointPair`, I would expect `close()` to close _that_ pair. But here `close()` closes the `clientEndpoints`, while the `pair` references the `workerEndpoints` 🤔

felixfbecker replied:
I guess it should really close both sides since a MessageChannel with only one side closed  doesn't really make sense

twop replied:
left only `frame.remove()`

felixfbecker commented:
@twop FYI I made the suggested improvements on a [separate branch](https://github.com/sourcegraph/sourcegraph/compare/direct-extension-host) and merged @lguychard's branch into it too. I'll continue to work on that given our resolution at web team sync

twop commented:
UPD:
- fixed tests, we got coverage from integration tests. It is awesome that we have those
- review feedback: callback -> Subscription, renames, deleted commented code

lguychard commented on shared/src/api/integration-test/testHelpers.ts:
```diff
@@ -91,7 +92,17 @@ export async function integrationTestContext(
     }
-    const client = await createExtensionHostClientConnection(clientEndpoints, services, initData, mocks)
+
+    // TODO (simon) do we need to handle unsub? I guess ports will be just garbage collected
```
```suggestion
```

I think not handling unsubscription is OK here
